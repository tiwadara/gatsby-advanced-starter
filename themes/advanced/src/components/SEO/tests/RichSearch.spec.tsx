import RichSearchTags, {
  getAuthorMetadata,
  generateOrganizationMetadata,
  generateArticleMetadata,
} from "../RichSearch";
import {
  tagListHasEmptyValues,
  sampleSeoData,
  containsEmptyValues,
} from "./TestUtils";

describe("seo module RichSearchTags", () => {
  it("generates correct tags on website pages", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoWebsite,
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });

    expect(generatedTags).toMatchSnapshot();
  });

  it("generates correct tags on article pages", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });

    expect(generatedTags).toMatchSnapshot();
  });

  it("doesn't generate tags when missing post coverImageUrl", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: { ...sampleSeoData.post, coverImageUrl: undefined },
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });
    expect(generatedTags).toStrictEqual([]);
  });

  it("doesn't generate tags when missing post description", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: { ...sampleSeoData.post, description: undefined },
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });

    expect(generatedTags).toStrictEqual([]);
  });

  it("generates correct tags when missing UserData", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: undefined,
      orgData: sampleSeoData.organization,
    });

    expect(generatedTags).toMatchSnapshot();
  });

  it("generates correct tags when missing OrgData", () => {
    expect.assertions(1);

    const generatedTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: sampleSeoData.user,
      orgData: undefined,
    });

    expect(generatedTags).toMatchSnapshot();
  });

  it("doesn't generate empty tags", () => {
    expect.assertions(4);

    // Test for website pages
    const generatedWebsiteTags = RichSearchTags({
      seoData: sampleSeoData.seoWebsite,
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });

    expect(tagListHasEmptyValues(generatedWebsiteTags)).toBe(false);

    // Test for article pages
    const generatedArticleTags = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: sampleSeoData.user,
      orgData: sampleSeoData.organization,
    });

    expect(tagListHasEmptyValues(generatedArticleTags)).toBe(false);

    // Test for missing UserData
    const generatedArticleTagsWithoutUserData = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: undefined,
      orgData: sampleSeoData.organization,
    });

    expect(tagListHasEmptyValues(generatedArticleTagsWithoutUserData)).toBe(
      false
    );

    // Test for missing OrgData
    const generatedTagsWithoutOrgData = RichSearchTags({
      seoData: sampleSeoData.seoArticle,
      postData: sampleSeoData.post,
      userData: sampleSeoData.user,
      orgData: undefined,
    });

    expect(tagListHasEmptyValues(generatedTagsWithoutOrgData)).toBe(false);
  });
});

describe("seo module RichSearchTags's utility getAuthorMetadata", () => {
  it("generates valid JSONLD", () => {
    expect.assertions(2);

    const generatedTags = getAuthorMetadata(sampleSeoData.user);

    expect(containsEmptyValues(generatedTags)).toBe(false);

    expect(() => JSON.stringify(generatedTags)).not.toThrow(Error);
  });
});

describe("seo module RichSearchTags's utility generateOrganizationMetadata", () => {
  it("generates valid JSONLD", () => {
    expect.assertions(2);

    const generatedTags = generateOrganizationMetadata(
      sampleSeoData.organization
    );

    expect(containsEmptyValues(generatedTags)).toBe(false);

    expect(() => JSON.stringify(generatedTags)).not.toThrow(Error);
  });
});

describe("seo module RichSearchTags's utility generateArticleMetadata", () => {
  it("generates valid JSONLD", () => {
    expect.assertions(2);

    const generatedTags = generateArticleMetadata(
      sampleSeoData.post,
      sampleSeoData.organization,
      sampleSeoData.user
    );

    expect(containsEmptyValues(generatedTags)).toBe(false);

    expect(() => JSON.stringify(generatedTags)).not.toThrow(Error);
  });
});
