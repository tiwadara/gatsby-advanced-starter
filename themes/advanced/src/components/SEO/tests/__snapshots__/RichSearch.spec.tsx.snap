// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seo module RichSearchTags generates correct tags on article pages 1`] = `
Array [
  <script
    type="application/ld+json"
  >
    {"@context":"http://schema.org","@type":"BlogPosting","image":"/logos/logo-1024.png","url":"/posts/url","headline":"Post title","name":"Post title","description":"Post description.","dateCreated":"2021-12-17T00:00:00.000Z","datePublished":"2021-12-17T00:00:00.000Z","dateModified":"2021-12-20T00:00:00.000Z","author":{"@type":"Person","givenName":"First","familyName":"Last","email":"AdvancedUser@example.com","address":"A full-stack web developer looking for a challenge!"},"creator":{"@type":"Person","givenName":"First","familyName":"Last","email":"AdvancedUser@example.com","address":"A full-stack web developer looking for a challenge!"},"publisher":{"@context":"https://schema.org","@type":"Organization","url":"https://gatsby-advanced-starter-demo.netlify.com","name":"Organization Name","description":"Organization description","logo":"/logos/logo-512.png"},"mainEntityOfPage":"True","keywords":["Tag1","Tag2"],"articleSection":"Tech","articleBody":"Post body"}
  </script>,
]
`;

exports[`seo module RichSearchTags generates correct tags on website pages 1`] = `
Array [
  <script
    type="application/ld+json"
  >
    {"@context":"https://schema.org","@type":"Organization","url":"https://gatsby-advanced-starter-demo.netlify.com","name":"Organization Name","description":"Organization description","logo":"/logos/logo-512.png"}
  </script>,
]
`;

exports[`seo module RichSearchTags generates correct tags when missing OrgData 1`] = `
Array [
  <script
    type="application/ld+json"
  >
    {"@context":"http://schema.org","@type":"BlogPosting","image":"/logos/logo-1024.png","url":"/posts/url","headline":"Post title","name":"Post title","description":"Post description.","dateCreated":"2021-12-17T00:00:00.000Z","datePublished":"2021-12-17T00:00:00.000Z","dateModified":"2021-12-20T00:00:00.000Z","author":{"@type":"Person","givenName":"First","familyName":"Last","email":"AdvancedUser@example.com","address":"A full-stack web developer looking for a challenge!"},"creator":{"@type":"Person","givenName":"First","familyName":"Last","email":"AdvancedUser@example.com","address":"A full-stack web developer looking for a challenge!"},"mainEntityOfPage":"True","keywords":["Tag1","Tag2"],"articleSection":"Tech","articleBody":"Post body"}
  </script>,
]
`;

exports[`seo module RichSearchTags generates correct tags when missing UserData 1`] = `
Array [
  <script
    type="application/ld+json"
  >
    {"@context":"http://schema.org","@type":"BlogPosting","image":"/logos/logo-1024.png","url":"/posts/url","headline":"Post title","name":"Post title","description":"Post description.","dateCreated":"2021-12-17T00:00:00.000Z","datePublished":"2021-12-17T00:00:00.000Z","dateModified":"2021-12-20T00:00:00.000Z","publisher":{"@context":"https://schema.org","@type":"Organization","url":"https://gatsby-advanced-starter-demo.netlify.com","name":"Organization Name","description":"Organization description","logo":"/logos/logo-512.png"},"mainEntityOfPage":"True","keywords":["Tag1","Tag2"],"articleSection":"Tech","articleBody":"Post body"}
  </script>,
]
`;
